File: app/node/math/math/math.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "math.h"

namespace olive
{

const QString MathNode::k_method_in = QStringLiteral("method_in");
const QString MathNode::k_param_a_in = QStringLiteral("param_a_in");
const QString MathNode::k_param_b_in = QStringLiteral("param_b_in");
const QString MathNode::k_param_c_in = QStringLiteral("param_c_in");

#define super MathNodeBase

MathNode::MathNode()
{
	add_input(k_method_in, NodeValue::k_combo,
			 InputFlags(k_input_flag_not_connectable | k_input_flag_not_keyframable));

	add_input(k_param_a_in, NodeValue::k_float, 0.0);
	set_input_property(k_param_a_in, QStringLiteral("decimalplaces"), 8);
	set_input_property(k_param_a_in, QStringLiteral("autotrim"), true);

	add_input(k_param_b_in, NodeValue::k_float, 0.0);
	set_input_property(k_param_b_in, QStringLiteral("decimalplaces"), 8);
	set_input_property(k_param_b_in, QStringLiteral("autotrim"), true);
}

QString MathNode::name() const
{
	// Default to naming after the operation
	if (parent()) {
		QString op_name = get_operation_name(get_operation());
		if (!op_name.isEmpty()) {
			return op_name;
		}
	}

	return tr("Math");
}

QString MathNode::id() const
{
	return QStringLiteral("org.olivevideoeditor.Olive.math");
}

QVector<Node::CategoryID> MathNode::category() const
{
	return { k_category_math };
}

QString MathNode::description() const
{
	return tr("Perform a mathematical operation between two values.");
}

void MathNode::retranslate()
{
	super::retranslate();

	set_input_name(k_method_in, tr("Method"));
	set_input_name(k_param_a_in, tr("Value"));
	set_input_name(k_param_b_in, tr("Value"));

	QStringList operations = { get_operation_name(k_op_add),
							   get_operation_name(k_op_subtract),
							   get_operation_name(k_op_multiply),
							   get_operation_name(k_op_divide),
							   get_operation_name(k_op_power) };

	set_combo_box_strings(k_method_in, operations);
}

ShaderCode MathNode::get_shader_code(const ShaderRequest &request) const
{
	return get_shader_code_internal(request.id, k_param_a_in, k_param_b_in);
}

void MathNode::value(const NodeValueRow &value, const NodeGlobals &globals,
					 NodeValueTable *table) const
{
	// Auto-detect what values to operate with
	// FIXME: Very inefficient
	NodeValueTable at, bt;
	at.push(value[k_param_a_in]);
	bt.push(value[k_param_b_in]);
	PairingCalculator calc(at, bt);

	// Do nothing if no pairing was found
	if (!calc.found_most_likely_pairing()) {
		return;
	}

	return value_internal(get_operation(), calc.get_most_likely_pairing(), k_param_a_in,
						 calc.get_most_likely_value_a(), k_param_b_in,
						 calc.get_most_likely_value_b(), globals, table);
}

void MathNode::process_samples(const NodeValueRow &values,
							  const SampleBuffer &input, SampleBuffer &output,
							  int index) const
{
	return process_samples_internal(values, get_operation(), k_param_a_in, k_param_b_in,
								  input, output, index);
}

}

File: src/node/rust/src/nodes/math.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Math node (C++ `src/node/src/math/math/math.{h,cpp}`,
//! `olive::MathNode`); behavior shared with other binary math nodes
//! lives in [`super::mathbase`] (C++ `MathNodeBase`).

use crate::factory::NodeMeta;
use crate::node::{Category, NodeBehavior, NodeCore};

/// Operation/method input id (C++ `k_method_in`). Type: combo;
/// flags: not-connectable, not-keyframable; combo strings are the
/// operation names Add/Subtract/Multiply/Divide/Power.
pub const METHOD_INPUT: &str = "method_in";

/// Operand A input id (C++ `k_param_a_in`). Type: float; default
/// `0.0`; properties: `decimalplaces = 8`, `autotrim = true`.
pub const PARAM_A_INPUT: &str = "param_a_in";

/// Operand B input id (C++ `k_param_b_in`). Type: float; default
/// `0.0`; properties: `decimalplaces = 8`, `autotrim = true`.
pub const PARAM_B_INPUT: &str = "param_b_in";

/// Operand C input id (C++ `k_param_c_in`). Declared as a static but
/// never added in the constructor — reserved/unused upstream.
pub const PARAM_C_INPUT: &str = "param_c_in";

/// Math node: applies a binary arithmetic operation to two values.
/// Unit-like — the C++ class has no own data members (inputs live in
/// [`NodeCore`], shared logic in [`super::mathbase`]).
pub struct MathNode;

impl MathNode {
	/// Construct the behavior struct (C++ `MathNode()`; the node's
	/// inputs are wired in [`create`]). Exposed so other nodes can own
	/// a child math node (e.g. `super::opacity` multiplies via one).
	pub fn new() -> Box<MathNode> {
		todo!()
	}
}

impl NodeBehavior for MathNode {
	/// Human-readable name (C++ `name()`): if the node is parented
	/// (i.e. owned as a child of another node) and the current
	/// operation has a name, returns that operation name
	/// (Add/Subtract/Multiply/Divide/Power); otherwise "Math".
	fn name(&self) -> &str {
		todo!()
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.math"
	}

	/// Categories (C++ `category()`).
	fn categories(&self) -> &[Category] {
		&[Category::Math]
	}

	/// Description (C++ `description()`).
	fn description(&self) -> &str {
		"Perform a mathematical operation between two values."
	}

	/// Localized input names (C++ `retranslate()`): `method_in` ->
	/// "Method", `param_a_in`/`param_b_in` -> "Value"; also sets the
	/// combo strings on `method_in` to the five operation names.
	fn input_name<'a>(&self, id: &'a str) -> &'a str {
		todo!()
	}

	/// Evaluate outputs (C++ `value()`): pushes both operands into
	/// single-value tables, runs the [`super::mathbase::PairingCalculator`]
	/// heuristic, and if a pairing was found delegates to
	/// [`super::mathbase::MathNodeBase::value_internal`] with the
	/// current operation; otherwise pushes nothing.
	fn value(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		time: oakcore_rs::Rational,
		table: &mut crate::value::NodeValueTable,
	) {
		todo!()
	}

	/// Process a span of samples (C++ `process_samples()`): delegates
	/// to [`super::mathbase::MathNodeBase::process_samples_internal`]
	/// with the current operation and the `param_a_in`/`param_b_in`
	/// ids (only used for the sample*number pairing).
	fn process_samples(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		range: oakcore_rs::TimeRange,
		output: &mut crate::value::SampleBuffer,
	) {
		todo!()
	}

	/// Shader code request (C++ `get_shader_code()`): delegates to
	/// [`super::mathbase::MathNodeBase::shader_code_internal`] with the
	/// request id (expected `"<op>.<pairing>.<type_a>.<type_b>"`) and
	/// the `param_a_in`/`param_b_in` uniform names.
	fn shader_code(&self, request: &str) -> Option<String> {
		todo!()
	}

	/// Deep copy (C++ `copy()`).
	fn duplicate(&self, core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		todo!()
	}
}

/// Constructor (C++ `MathNode::MathNode()`): adds `method_in` as a
/// not-connectable/not-keyframable combo, and `param_a_in`/
/// `param_b_in` as float inputs defaulting to 0.0 with
/// `decimalplaces = 8` and `autotrim = true`.
pub fn create() -> (NodeCore, Box<dyn NodeBehavior>) {
	todo!()
}

/// Register this node type (C++ `k_math_node` in
/// `factory.cpp::create_from_factory_index`).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.math",
		name: "Math",
		categories: &[Category::Math],
		create,
	});
}
